### PR TITLE
only load a stdlib-named C extension from the standard library

### PR DESCRIPTION
**Pull request by kali834x**
## Type of Changes

|     | Type          |
| --- | ------------- |
| ✓   | :bug: Bug fix |

## Description

`_can_load_extension` only receives the module name, and `is_stdlib_module` is a
plain name test, so any C extension whose name matches a stdlib module is
imported no matter where it came from. `ImportlibFinder.find_module` walks
`sys.path` in order and tries the extension suffixes before `.py`, and the
analysed project's directory comes first, so a `colorsys.cpython-313-x86_64-linux-gnu.so`
committed to a repository resolves ahead of the real module and
`ast_from_module_name` hands it to `load_module_from_name`, which runs its
module init. That happens with `always_load_extensions` off and an empty
`extension-pkg-allow-list`, so the opt-in that guards extension loading is
skipped entirely and linting a checkout executes code out of it. Pass the
resolved location down and require it to be inside `STD_LIB_DIRS` before the
stdlib name is trusted, excluding `EXT_LIB_DIRS` because site-packages sits
below the stdlib directory in a virtualenv. Extensions shipped with the
interpreter still load from `lib-dynload`, and an explicit allow-list entry
still wins.

### CODE REVIEW

**Review by DanielNoord — COMMENTED** [review]
I like it!

@jacobtylerwalls generally is quite knowledgeable about our handling of imports. Do you have time for a review as well?

**Review by yangfan-yf-yf — CHANGES_REQUESTED** [review]
The location check is the right boundary, but the regression currently only exercises a temporary directory outside both library roots. It does not cover the security-relevant branch that rejects `EXT_LIB_DIRS` before accepting a path under `STD_LIB_DIRS`.

In a virtual environment, `site-packages` can sit below the standard-library directory, which is exactly why the early `EXT_LIB_DIRS` rejection is needed. Please add a focused regression for a location under `EXT_LIB_DIRS` (for example, an extension-suffixed `colorsys` path) and assert that it is not accepted as a standard-library location. That both guards the virtualenv case and covers the currently missed branch.

I verified locally that the proposed predicate rejects such a path, and the existing temporary-directory regression passes.

**Comment by kali834x** [comment]
Good point, that branch was untested. Added `IsStdLibPathTest` in tests/test_modutils.py: it builds a `colorsys` extension path under each `EXT_LIB_DIRS` entry and asserts `is_stdlib_path` rejects it, plus a positive assertion on a real stdlib file so the check isn't trivially true.

Confirmed it fails if the `EXT_LIB_DIRS` rejection is dropped (site-packages resolves under the stdlib dir here), so the virtualenv case is covered now.

**Comment by yangfan-yf-yf** [comment]
Thanks for adding both requested boundary tests. I rechecked the current head: the focused modutils/manager selection passes (`3 passed`), Ruff passes, and the hosted checks have no failures.

The branch is currently behind `main` and GitHub reports it as conflicting. The only merge conflict I find against current `main` is in `ChangeLog`; the product and test files merge cleanly. Could you update the branch and resolve that changelog conflict? I will re-review the resulting head afterward.

**Review by DanielNoord — COMMENTED** [review]
Needs a rebase :)

**Comment by kali834x** [comment]
Rebased onto main and resolved the ChangeLog conflict. The new entry now sits under the 4.4.0 section alongside the data-descriptor fix that landed there. Product and test files were unchanged by the rebase, and the focused modutils/manager tests still pass.

**Comment by jacobtylerwalls** [comment]
>  Do you have time for a review as well?

I can try over the next few days, sorry!

File: astroid/manager.py
Comment (jacobtylerwalls) [inline astroid/manager.py]: We need a test case for this `try` branch where `found_spec.type == spec.ModuleType.C_EXTENSION`

File: astroid/manager.py [hunk 1]
```diff
@@ -183,10 +184,15 @@ def _build_namespace_module(
...
 
-    def _can_load_extension(self, modname: str) -> bool:
+    def _can_load_extension(self, modname: str, location: str | None) -> bool:
         if self.always_load_extensions:
             return True
-        if is_stdlib_module(modname):
+        # Only trust a stdlib name that resolves to the stdlib itself.
+        if (
+            is_stdlib_module(modname)
+            and location is not None
```
Comment (jacobtylerwalls) [inline astroid/manager.py:193]: Nothing fails when I remove `and location is not None`, so we must be missing a test case for when location is `None`, e.g. for a stdlib c module like `time`.

File: tests/test_manager.py [hunk 2]
```diff
@@ -146,6 +148,21 @@ def test_ast_from_module_name_astro_builder_exception(self) -> None:
             "unhandledModule",
         )
 
+    def test_ast_from_module_name_extension_shadowing_stdlib(self) -> None:
+        """An extension outside the stdlib is stubbed, not imported."""
+        modname = "colorsys"
```
Comment (jacobtylerwalls) [inline tests/test_manager.py:153]: This test fails when I change this to `time`, although it's not clear to me if if the shadowing file is the one that's interpreted by astroid or not.

Can you walk me through this?

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for working on this.

File: astroid/manager.py (hunk 1 above)
Comment (kali834x) [reply astroid/manager.py:193]: Right, that branch was uncovered. The reason nothing fails with a real module is that `time` is a statically linked builtin (C_BUILTIN, location None), so it never reaches `_can_load_extension`, which is only called for C_EXTENSION. The guard is there because `found_spec.location` is typed `str | None` and `is_stdlib_path` calls `os.path.realpath`, which raises on None.

Pushed a focused test that calls `_can_load_extension("colorsys", None)` directly and asserts False. Drop the guard and it errors with TypeError instead of returning False, so the branch is pinned now.

File: tests/test_manager.py (hunk 2 above)
Comment (kali834x) [reply tests/test_manager.py:153]: Good question, and the difference is the point of picking `colorsys` here. `colorsys` is a pure-Python stdlib module, so a `colorsys.<ext>.so` planted at the front of `sys.path` resolves ahead of the real `colorsys.py`. astroid sees a C_EXTENSION whose location is the planted file, and that is the file it would import, which is the bug. The fix stubs it instead.

`time` behaves differently because it is a statically linked builtin. `BuiltinImporter` sits ahead of the path-based finders, so it resolves `time` before `sys.path` is ever scanned, and the planted `time.so` is ignored. astroid gets a C_BUILTIN with location None, never reaches the extension gate, and imports the real builtin, so nothing is stubbed and the assertion fails. The shadowing file is not what astroid interprets for `time`, which is why the test uses a pure-Python stdlib name.

**Review by jacobtylerwalls — DISMISSED** [review]
LGTM!

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Let's fix pre-commit and then merge :)

**Comment by kali834x** [comment]
The pre-commit failure was the merge conflict: main moved the changelog to towncrier since the last rebase, so pre-commit.ci couldn't compute a merge. Rebased again and moved the entry from ChangeLog into doc/whatsnew/fragments/3193.bugfix. pre-commit passes locally on the diff, no other changes.

**Review by yangfan-yf-yf — APPROVED** [review]
I rechecked c9124dad after both rebases. The requested `EXT_LIB_DIRS` regression, real-stdlib positive control, and `None`-location regression are now present. All four focused tests pass locally. Black, the news-fragment check, Ruff, mypy, Pylint, and diff validation pass, and all hosted checks are green. Current `main` only changes three configuration files and does not overlap this patch. I found no remaining blocker.

**Review by Pierre-Sassoulas — APPROVED** [review]
